sdk/nodejs/cmd/run/run.ts

```diff
@@ -251,6 +256,28 @@ ${defaultMessage}`);
         // Now go ahead and execute the code. The process will remain alive until the message loop empties.
         log.debug(`Running program '${program}' in pwd '${process.cwd()}' w/ args: ${programArgs}`);
         try {
+            const packageObject = packageObjectFromProjectRoot(projectRootFromProgramPath(program));
+
+            // We use dynamic import instead of require for projects using native ES modules instead of commonjs
+            if (packageObject["type"] === "module") {
+                // Use the same behaviour for loading the main entrypoint as `node <program>`.
+                // See https://github.com/nodejs/node/blob/master/lib/internal/modules/run_main.js#L74.
+                const mainPath: string = require("module").Module._findPath(path.resolve(program), null, true) || program;
+                const main = path.isAbsolute(mainPath) ? url.pathToFileURL(mainPath).href : mainPath;
+                // Workaround for typescript transpiling dynamic import into `Promise.resolve().then(() => require`
+                // Follow this issue for progress on when we can remove this:
+                // https://github.com/microsoft/TypeScript/issues/43329
+                //
+                // Workaround inspired by es-module-shims:
+                // https://github.com/guybedford/es-module-shims/blob/main/src/common.js#L21
+                // eslint-disable-next-line no-eval
+                const dynamicImport = (0, eval)("u=>import(u)");
```

Comment:
Some new developments since I opened the original PR: 

According to https://github.com/microsoft/TypeScript/issues/43329#issuecomment-926931937, we should be able to rid of this workaround and use `import(main)` directly if we upgraded to TypeScript 4.5 and used `module: node12`.

Reply:
> if we upgraded to TypeScript 4.5 and used module: node12, which preserves dynamic imports instead of transpiling them into requires.

I believe `module: node12` was removed from TypeScript 4.5, and is only available in nightlies: https://github.com/microsoft/TypeScript/issues/46454.  I'm also not too worried about the workaround you used here, and shifting to a newer TypeScript version and especially newer approach to compilation of modules for this library is likely to introduce it's own interesting side effects.  I'm happy to wait on that to handle separately, most likely after a TypeScript release with this new `module` flag is released.